modules/local/collect_stats.nf

```diff
@@ -21,114 +21,76 @@ process COLLECT_STATS {
-                }
-            )
+        read_trimlogs = """
+    read_tsv(
+        pipe("grep -H 'Reads written (passing filters)' *trimming_report.txt"),
+        col_names = c('s'), col_types = 'c'
+    ) %>%
+        transmute(
+            sample    = str_replace(s, '^(.+)_\\\\d\\\\.fastq.*', '\\\\1'),
+            n_trimmed = str_replace(s, '.* (\\\\d+) .*', '\\\\1') %>% as.integer() * 2
         ) %>%
-        unnest(d) %>%
-        rename(n_trimmed = V1) %>%
-        mutate(n_trimmed = n_trimmed*2) %>%
+        pivot_longer(2:ncol(.), names_to = 'm', values_to = 'v') %>%
+        """
+    }
+
+    def read_bbduklogs = "tibble(sample = character(), m = character(), v = numeric())"
+    if ( bbduklogs ) {
+        read_bbduklogs = """
+        read_fwf(pipe("grep -H 'Result:' *.bbduk.log"), fwf_widths(c(1e5), c('c')), col_types = 'c') %>%
+            transmute(
+                sample = str_remove(c, '.bbduk.log.*'),
+                m = 'n_non_contaminated',
+                v = str_replace(c, '.*Result:\\\\D*(\\\\d+).*', '\\\\1') %>% as.numeric()
+            )
         """
-    } else {
-        read_trimlogs = "%>%"
     }
 
     """
     #!/usr/bin/env Rscript
 
-    library(data.table)
-    library(dtplyr)
     library(dplyr)
     library(readr)
     library(purrr)
     library(tidyr)
     library(stringr)
 
-    TYPE_ORDER = c('n_trimmed', 'n_non_contaminated', 'idxs_n_mapped', 'idxs_n_unmapped', 'n_feature_count', 'n_feature_cds')
+    TYPE_ORDER = c('n_trimmed', 'n_non_contaminated', 'idxs_n_mapped', 'idxs_n_unmapped', 'CDS', 'rRNA', 'tRNA', 'tmRNA')
 
-    # Collect stats for each sample, create a table in long format that can be appended to
-    t <- tibble(sample = c("${samples.join('", "')}")) ${read_trimlogs}
+    t <- ${read_trimlogs}
         # add samtools idxstats output
-        mutate(
-            i = map(
-                sample,
-                function(s) {
-                    fread(
-                        cmd = sprintf("grep -v '^*' %s*idxstats", s),
-                        sep = '\\t',
-                        col.names = c('chr', 'length', 'idxs_n_mapped', 'idxs_n_unmapped')
-                        ) %>%
-                        lazy_dt() %>%
-                        summarise(idxs_n_mapped = sum(idxs_n_mapped), idxs_n_unmapped = sum(idxs_n_unmapped)) %>%
-                        as_tibble()
-                }
-            )
-        ) %>%
-        unnest(i) %>%
-        pivot_longer(2:ncol(.), names_to = 'm', values_to = 'v') %>%
         union(
-            # Total observation after featureCounts
-            tibble(file = Sys.glob('*.counts.tsv.gz')) %>%
-                mutate(
-                    d = map(
-                        file,
-                        function(f) fread(cmd = sprintf("gunzip -c %s", f),
-                        sep = '\\t'
-                        )
-                    )
+            read_tsv(
+                pipe("grep -vH '*' *.idxstats"),
```

Comment:
I never saw pipe function. does it work as substitute of sprintf or what?

Reply:
It works like a pipe, the same way that `cmd` in `readf` does: Read the output of a Unix command as if it was a file. I didn't realize this until quite recently.